builder/upcloud/driver.go

```diff
@@ -0,0 +1,299 @@
+	response, err := d.svc.GetStorageDetails(&request.GetStorageDetailsRequest{
+		UUID: storageUuid,
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("Error fetching storages: %s", err)
+	}
+	return &response.Storage, nil
+}
+
+func (d *driver) getStorageByName(storageName string) (*upcloud.Storage, error) {
+	response, err := d.svc.GetStorages(&request.GetStoragesRequest{
+		Type: upcloud.StorageTypeTemplate,
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("Error fetching storages: %s", err)
+	}
+
+	var found bool
+	var storage upcloud.Storage
+	for _, s := range response.Storages {
+		if strings.Contains(strings.ToLower(s.Title), strings.ToLower(storageName)) {
+			found = true
+			storage = s
+			break
+		}
+	}
+
+	if !found {
+		return nil, fmt.Errorf("Failed to find storage by name %q", storageName)
+	}
+	return &storage, nil
+}
+
+func (d *driver) waitDesiredState(serverUuid string, state string) error {
+	request := &request.WaitForServerStateRequest{
+		UUID:         serverUuid,
+		DesiredState: state,
+		Timeout:      d.config.Timeout,
+	}
+	if _, err := d.svc.WaitForServerState(request); err != nil {
+		return fmt.Errorf("Error while waiting for server to change state to %q: %s", state, err)
+	}
+	return nil
+}
+
+func (d *driver) waitUndesiredState(serverUuid string, state string) error {
+	request := &request.WaitForServerStateRequest{
+		UUID:           serverUuid,
+		UndesiredState: state,
+		Timeout:        d.config.Timeout,
+	}
+	if _, err := d.svc.WaitForServerState(request); err != nil {
+		return fmt.Errorf("Error while waiting for server to change state from %q: %s", state, err)
+	}
+	return nil
+}
+
+func (d *driver) getServerDetails(serverUuid string) (*upcloud.ServerDetails, error) {
+	response, err := d.svc.GetServerDetails(&request.GetServerDetailsRequest{
+		UUID: serverUuid,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get details for server: %s", err)
+	}
+	return response, nil
+}
+
+func (d *driver) getServerStorage(serverUuid string) (*upcloud.ServerStorageDevice, error) {
+	details, err := d.getServerDetails(serverUuid)
+	if err != nil {
+		return nil, err
+	}
+
+	var found bool
+	var storage upcloud.ServerStorageDevice
+	for _, s := range details.StorageDevices {
+		if s.Type == upcloud.StorageTypeDisk {
+			found = true
+			storage = s
+			break
+		}
+	}
+	if !found {
+		return nil, fmt.Errorf("Failed to find storage type disk for server %q", serverUuid)
+	}
+	return &storage, nil
+}
+
+func (d *driver) prepareCreateRequest(storageUuid, sshKeyPublic string) *request.CreateServerRequest {
+	title := fmt.Sprintf("packer-%s-%s", d.config.TemplatePrefix, GetNowString())
+	hostname := d.config.TemplatePrefix
+	titleDisk := fmt.Sprintf("%s-disk1", title)
+
+	return &request.CreateServerRequest{
+		Title:            title,
+		Hostname:         hostname,
+		Zone:             d.config.Zone,
+		PasswordDelivery: request.PasswordDeliveryNone,
+		CoreNumber:       2,
+		MemoryAmount:     2048,
```

Comment:
Rather use some plan instead (`GET https://api.upcloud.com/1.3/plan`), passed with `Plan`

Reply:
as the customer will be billed for the first hour when spinning up a server with this